core/arch/atomic_exports.h

```diff
@@ -684,23 +728,28 @@ atomic_dec_becomes_zero(volatile int *var)
             ATOMIC_ADD_EXCHANGE_suffix("", var, val, res)
 #        define ATOMIC_ADD_EXCHANGE_int64(var, val, res) \
             ATOMIC_ADD_EXCHANGE_suffix("d", var, val, res)
-#        define ATOMIC_COMPARE_EXCHANGE_suffix(suffix, var, compare, exchange)           \
-            __asm__ __volatile__("   dmb ish                      \n\t"                  \
-                                 "2: ldrex" suffix " r2, %0       \n\t"                  \
-                                 "   cmp" suffix " r2, %1       \n\t"                    \
-                                 "   bne    1f                    \n\t"                  \
-                                 "   strex" suffix " r3, %2, %0   \n\t"                  \
-                                 "   cmp   r3, #0                 \n\t"                  \
-                                 "   bne   2b                     \n\t"                  \
-                                 "   cmp" suffix " r2, %1       \n\t" /* for SET_FLAG */ \
-                                 "1: clrex                        \n\t"                  \
-                                 : "=Q"(var) /* no offset for ARM mode */                \
-                                 : "r"(compare), "r"(exchange)                           \
+/* "matched" is assumed to be a single byte bool. */
+#        define ATOMIC_COMPARE_EXCHANGE_suffix(suffix, var, compare, exchange, matched) \
+            __asm__ __volatile__("   dmb ish                      \n\t"                 \
+                                 "2: ldrex" suffix " r2, %0       \n\t"                 \
+                                 "   cmp" suffix " r2, %2         \n\t"                 \
+                                 "   bne    1f                    \n\t"                 \
+                                 "   strex" suffix " r3, %3, %0   \n\t"                 \
+                                 "   cmp   r3, #0                 \n\t"                 \
+                                 "   bne   2b                     \n\t"                 \
+                                 "   cmp" suffix " r2, %2         \n\t"                 \
```

Comment:
If control reaches L740, isn't r2 == %2 guaranteed? I guess this is just to set the conditional flags for the beq at L743 which is shared between two control paths. Could be simpler to just branch to 3 here? And remove the mov and beq at L742 and L743?

Reply:
https://developer.arm.com/documentation/dui0473/j/arm-and-thumb-instructions/clrex says

```
CLREX returns a closely-coupled exclusive access monitor to its open-access state.
This removes the requirement for a dummy store to memory.
```

Reply:
It executed clrex before on the success path, so I would prefer to not change something like that. If we want to change that I would prefer to do it separately.

Reply:
Maybe add an XXX comment?

Reply:
@egrimley-arm I believe you authored DEF_atomic_compare_exchange at https://github.com/DynamoRIO/dynamorio/blob/master/core/arch/atomic_exports.h#L546: is there a reason to have CLREX on the success path or it has no benefit and it's just that a branch to skip it would cost more than executing it?

Reply:
@egrimley @jackgallagher-arm any idea on this CLREX question?